samples/core/multi-device/main.c

```diff
@@ -0,0 +1,646 @@
+    {
+        printf("Creating sub-devices...");
+        fflush(stdout);
+    }
+
+#if CL_HPP_TARGET_OPENCL_VERSION < 120
+    fprintf(stderr,
+            "Error: OpenCL subdevices not supported before version 1.2 ");
+    exit(EXIT_FAILURE);
+#endif
+
+    // Create subdevices, each with half of the compute units available.
+    cl_uint max_compute_units = 0;
+    OCLERROR_RET(clGetDeviceInfo(dev, CL_DEVICE_MAX_COMPUTE_UNITS,
+                                 sizeof(cl_uint), &max_compute_units, NULL),
+                 error, dev);
+    cl_device_partition_property subdevices_properties[] = {
+        (cl_device_partition_property)CL_DEVICE_PARTITION_EQUALLY,
+        (cl_device_partition_property)(max_compute_units / 2), 0
+    };
+
+    // Initialize subdevices array with one device and then reallocate for
+    // MacOS and Windows not to complain about NULL subdevices array.
+    cl_uint subdev_count = 1;
+    cl_device_id* subdevices =
+        (cl_device_id*)malloc(subdev_count * sizeof(cl_device_id));
+
+    OCLERROR_RET(clCreateSubDevices(dev, subdevices_properties,
+                                    max_compute_units, subdevices,
+                                    &subdev_count),
+                 error, dev);
+
+    if (subdev_count < 2)
+    {
+        fprintf(stderr, "Error: OpenCL cannot create subdevices");
+        exit(EXIT_FAILURE);
+    }
+
+    subdevices =
+        (cl_device_id*)realloc(subdevices, subdev_count * sizeof(cl_device_id));
+    OCLERROR_RET(clCreateSubDevices(dev, subdevices_properties, subdev_count,
+                                    subdevices, NULL),
+                 error, subdevs);
+
+    OCLERROR_PAR(context = clCreateContext(NULL, subdev_count, subdevices, NULL,
+                                           NULL, &error),
+                 error, subdevs);
+
+    // Read kernel file.
+    const char* kernel_location = "./convolution.cl";
+    char *kernel = NULL, *tmp = NULL;
+    size_t program_size = 0;
+    OCLERROR_PAR(
+        kernel = cl_util_read_text_file(kernel_location, &program_size, &error),
+        error, contx);
+    MEM_CHECK(tmp = (char*)realloc(kernel, program_size), error, ker);
+    kernel = tmp;
+
+    // Compile kernel.
+    if (diag_opts.verbose)
+    {
+        printf("done.\nCompiling kernel...");
+        fflush(stdout);
+    }
+    OCLERROR_PAR(program = clCreateProgramWithSource(
+                     context, 1, (const char**)&kernel, &program_size, &error),
+                 error, ker);
+
+    // If no -cl-std option is specified then the highest 1.x version
+    // supported by each device is used to compile the program. Therefore,
+    // it's only necessary to add the -cl-std option for 2.0 and 3.0 OpenCL
+    // versions.
+    char compiler_options[1023] = "";
+#if CL_HPP_TARGET_OPENCL_VERSION >= 300
+    strcat(compiler_options, "-cl-std=CL3.0 ");
+#elif CL_HPP_TARGET_OPENCL_VERSION >= 200
+    strcat(compiler_options, "-cl-std=CL2.0 ");
+#endif
```

Comment:
I'll answer my own question - it looks like the kernel uses the loop unroll attribute, so it would be beneficial to compile for a higher OpenCL version.

Maybe we need an OpenCL SDK utility to get a build option representing the highest OpenCL C version supported by a device?

Or, we just don't worry about the loop unroll attribute, simplify this sample, and compiler for OpenCL 1.x unconditionally.

Reply:
Same as before, we just do this so the program logic does not depend on the version with which it is compiled.
Regarding the version that we currently use for compiling, I would keep the 3.0 and the unroll, but I also don't have strong feelings about this so we could as well just drop it to the minimum needed (1.2).

Reply:
I have slight preference towards keeping the sample simple and removing this logic, but I'm fine keeping it if you'd like.  If we do keep it though, please make it a runtime check based on the properties of the device, specifically whether it supports OpenCL C 3.0 or OpenCL 2.0, and not a compile-time check based on a C++ header define.

Reply:
Runtime check added!